sklearn/linear_model/sag_fast.pyx

```diff
@@ -496,22 +563,66 @@ cdef double scale_weights(double* weights, double wscale, int n_features,
     coefficients and reset the just-in-time update system.
     This also limits the size of `cumulative_sums`.
     """
-    cdef int feature_ind, class_ind, idx
+    cdef int feature_ind, class_ind, idx, f_idx, lagged_ind, last_update_ind
     cdef double cum_sum
-    idx = -1
-    for feature_ind in range(n_features):
-        cum_sum = cumulative_sums[sample_itr]
+    for feature_ind in range(xnnz):
+        if not reset:
+            feature_ind = x_ind_ptr[feature_ind]
+        f_idx = feature_ind * n_classes
+
+        cum_sum = cumulative_sums[sample_itr - 1]
+        if prox:
+            cum_sum_prox = cumulative_sums_prox[sample_itr - 1]
         if feature_hist[feature_ind] != 0:
             cum_sum -= cumulative_sums[feature_hist[feature_ind] - 1]
-
-        for class_ind in range(n_classes):
-            idx += 1  # idx = feature_ind * n_classes + class_ind
-            weights[idx] -= cum_sum * sum_gradient[idx]
-            weights[idx] *= wscale
-
-        feature_hist[feature_ind] = (sample_itr + 1) % n_samples
-
-    cumulative_sums[sample_itr % n_samples] = 0.0
+            if prox:
+                cum_sum_prox -= cumulative_sums_prox[feature_hist[feature_ind] - 1]
+        if not prox:
+            for class_ind in range(n_classes):
+                idx = f_idx + class_ind
+                weights[idx] -= cum_sum * sum_gradient[idx]
+                if reset:
+                    weights[idx] *= wscale
+                    if not skl_isfinite(weights[idx]):
+                        with gil:
+                            raise_infinite_error(n_iter)
+        else:
+            for class_ind in range(n_classes):
+                idx = f_idx + class_ind
+                if fabs(sum_gradient[idx] * cum_sum) < cum_sum_prox:
```

Comment:
Even without the full blog post with the derivation it would be great to have an inline comment to state:
- what you said in https://github.com/scikit-learn/scikit-learn/pull/8446/files#r103928318
- give credits to the lightning implementation